Adding function to calculate HF orbitals and made minor reformat - Ferminet

shaipranesh2 opened the pull request:
## Description
Added prepare_hf_solution function to calculate hartree fock solution required for pertaining.
Reformatted the code into Ferminet class (class where ferminet model will implemented) and FerminetModel class (driver class providing all the necessary data to Ferminet class)

Fix #(issue)
Fixed the calculation of the spin of the molecule system

<!-- Please include a summary of the change and which issue is fixed.
Please also include relevant motivation and context.
List any dependencies that are required for this change. -->


## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [ ] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [ ] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.32.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [ ] Run `flake8 <modified file> --count` and check no errors
  - [ ] Run `python -m doctest <modified file>` and check no errors
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

rbharath commented on deepchem/models/tests/test_ferminet.py:
```diff
@@ -5,60 +5,36 @@
 
 
 @pytest.mark.torch
-def test_prepare_input_stream():
```
Why are we deleting this older test?

shaipranesh2 replied:
The old test checks for input streams shape, which is nothing but each electrons distance from each nuclei, this can be done using deepchem utils pairwise distance itself and doesn't require tests!

rbharath commented on deepchem/models/torch_models/ferminet.py:
```diff
@@ -22,9 +24,55 @@ def test_f(x: np.ndarray) -> np.ndarray:
 
 
-class Ferminet:
+class Ferminet(torch.nn.Module):
```
Can you add this new layer to torch_layers.csv and layers.rst in the docs/?

shaipranesh2 replied:
Can I do this once the model is fully completed? As I am breaking the chunk of PR into smaller ones and am adding one function by function.

rbharath commented on deepchem/models/torch_models/ferminet.py:
```diff
@@ -22,9 +24,55 @@ def test_f(x: np.ndarray) -> np.ndarray:
 
-class Ferminet:
+class Ferminet(torch.nn.Module):
+    """Approximates the log probability of the wave function of a molecule system using DNNs.
```
Can you add a usage example here?

shaipranesh2 replied:
Can I do this once the model is fully completed? As I am breaking the chunk of PR into smaller ones and am adding one function by function.

rbharath commented on deepchem/models/torch_models/ferminet.py:
```diff
@@ -22,9 +24,55 @@ def test_f(x: np.ndarray) -> np.ndarray:
+        self.determinant = determinant
+
+
+class FerminetModel(TorchModel):
```
Can you add this class to the docs?

shaipranesh2 replied:
Can I do this once the model is fully completed? As I am breaking the chunk of PR into smaller ones and am adding one function by function.

rbharath commented on deepchem/models/torch_models/ferminet.py:
```diff
@@ -22,9 +24,55 @@ def test_f(x: np.ndarray) -> np.ndarray:
+    solution of a many-electron system.
 
     This model aims to calculate the ground state energy of a multi-electron system
     using a baseline solution as the Hartree-Fock. An MCMC technique is used to sample
```
Can you format the reference to be in a References section for numpydoc? Can you also add a usage example?

shaipranesh2 replied:
done

rbharath commented on deepchem/models/torch_models/ferminet.py:
```diff
@@ -42,48 +90,41 @@ class Ferminet:
         self.spin = spin
-        self.ion_charge = charge
-
-    def prepare_input_stream(self,) -> Tuple[Any, Any, Any, Any]:
```
Why are we deleting this helper function?

shaipranesh2 replied:
This helper function computes pairwise distances of each electron from nuclei and can be done in forward itself

rbharath commented on deepchem/models/torch_models/ferminet.py:
```diff
@@ -124,46 +164,77 @@ def prepare_input_stream(self,) -> Tuple[Any, Any, Any, Any]:
         total_electrons = np.sum(self.electron_no)
-        self.up_spin = (total_electrons + 2 * self.spin) // 2
-        self.down_spin = (total_electrons - 2 * self.spin) // 2
+
```
Could you give me a summary of the refactoring changes you've made here?

shaipranesh2 replied:
The changes I have done are:
1. added a new helper function prepare_hf which gives hartree fock orbitals values for pretraining.
2. Deleted the code to prepare input streams as it should be done in forwards, as we need the gradient of output wrt to the electron's coordinates we are going to pass

For now, this model is empty and have added the helper function only. In successive PRs I will complete the forward function and add  example

rbharath reviewed (commented):
Good start. I have some questions and requests for more docs below

shaipranesh2 commented:
@rbharath I have reformatted to numpydocs style.
As this model is not complete now, I thought of not adding it to the layers.rst and in other docs.
I have added a helper function for pertaining and removed unnecessary functions for calculating input streams.
In the next PR I will be adding forward function of the model!

Is this fine?

rbharath reviewed (approved):
LGTM
